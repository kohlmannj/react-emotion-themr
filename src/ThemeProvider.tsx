import React, { FunctionComponent } from 'react';
import { ThemeProvider as CssThemeProvider } from '@friendsofreactjs/react-css-themr';
import {
  createEmotionContextTheme,
  TReactEmotionThemrContextTheme,
} from './createEmotionContextTheme';

export interface ThemeProviderProps {
  innerRef?: Function;
  theme: TReactEmotionThemrContextTheme;
}

export const ThemeProvider: FunctionComponent<ThemeProviderProps> = ({
  children,
  innerRef,
  theme,
}) => (
  <CssThemeProvider theme={createEmotionContextTheme(theme)} innerRef={innerRef}>
    {children}
  </CssThemeProvider>
);
