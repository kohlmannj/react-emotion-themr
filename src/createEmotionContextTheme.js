export function createEmotionContextTheme(contextThemeObj) {
  if (typeof contextThemeObj !== 'object' || contextThemeObj === null) {
    return contextThemeObj;
  }

  return Object.keys(contextThemeObj).reduce(
    (emotionContextTheme, componentName) => ({
      ...emotionContextTheme,
      [componentName]: createEmotionCssModule(contextThemeObj[componentName]),
    }),
    {}
  );
}
