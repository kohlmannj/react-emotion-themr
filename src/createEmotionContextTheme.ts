import { TReactCSSThemrTheme } from '@friendsofreactjs/react-css-themr';
import {
  createEmotionTheme,
  TReactEmotionThemrTheme,
  RenderedEmotionTheme,
} from './createEmotionTheme';

export type TReactEmotionThemrContextTheme = {
  [key: string]: TReactEmotionThemrTheme;
};

export function createEmotionContextTheme<T>(
  contextThemeObj: T
): RenderedEmotionTheme<T, TReactEmotionThemrContextTheme> {
  if (
    typeof contextThemeObj === 'object' &&
    typeof contextThemeObj !== 'undefined' &&
    contextThemeObj !== null
  ) {
    return Object.keys(contextThemeObj).reduce(
      (emotionContextTheme, componentName) => ({
        ...emotionContextTheme,
        [componentName]: createEmotionTheme(
          (contextThemeObj as TReactEmotionThemrTheme)[componentName]
        ),
      }),
      {} as TReactCSSThemrTheme
    ) as T extends TReactEmotionThemrContextTheme ? TReactCSSThemrTheme : never;
  }

  return contextThemeObj as T extends TReactEmotionThemrContextTheme ? never : T;
}
