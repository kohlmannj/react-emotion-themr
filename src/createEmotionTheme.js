import { css } from 'emotion';

export function createEmotionTheme(classNamesObj) {
  if (typeof classNamesObj !== 'object' || classNamesObj === null) {
    return classNamesObj;
  }

  return Object.keys(classNamesObj).reduce((emotionTheme, className) => {
    const value = classNamesObj[className];
    const typeofValueIsNonNullObject = typeof value === 'object' && value !== null;

    // Only add the className if `value` is a non-null object or a string
    if (typeofValueIsNonNullObject || typeof value === 'string') {
      return {
        ...emotionTheme,
        [className]: typeofValueIsNonNullObject ? css(value, `label:${className};`) : value,
      };
    }

    return emotionTheme;
  }, {});
}
