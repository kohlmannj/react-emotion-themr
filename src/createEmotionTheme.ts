import { TReactCSSThemrTheme } from '@friendsofreactjs/react-css-themr';
import { CSSObject } from 'create-emotion';
import { css } from 'emotion';

export type TReactEmotionThemrTheme = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: CSSObject | any;
};

export type RenderedEmotionTheme<T, V = TReactEmotionThemrTheme> = T extends V
  ? TReactCSSThemrTheme
  : T;

export function createEmotionTheme<T>(classNamesObj: T): RenderedEmotionTheme<T> {
  if (
    typeof classNamesObj === 'object' &&
    typeof classNamesObj !== 'undefined' &&
    classNamesObj !== null
  ) {
    return Object.keys(classNamesObj).reduce(
      (emotionTheme, className) => {
        const value = (classNamesObj as TReactEmotionThemrTheme)[className];

        if ((typeof value === 'object' && value !== null) || typeof value === 'string') {
          return {
            ...emotionTheme,
            [className]:
              typeof value === 'object' && value !== null
                ? css(value, `label:${className};`)
                : value,
          } as TReactCSSThemrTheme;
        }

        return emotionTheme;
      },
      {} as TReactCSSThemrTheme
      /** @see https://github.com/microsoft/TypeScript/issues/22735#issuecomment-376960435 */
    ) as T extends TReactEmotionThemrTheme ? TReactCSSThemrTheme : never;
  }

  return classNamesObj as T extends TReactEmotionThemrTheme ? never : T;
}
