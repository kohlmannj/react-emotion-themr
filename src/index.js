export * from './createEmotionContextTheme';
export * from './createEmotionTheme';
export * from './ThemeProvider';
export * from './themr';
