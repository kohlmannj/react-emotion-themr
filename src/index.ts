export * from './createEmotionContextTheme';
export * from './createEmotionTheme';
export * from './ThemeProvider';
export * from './themeable';
export * from './themr';
