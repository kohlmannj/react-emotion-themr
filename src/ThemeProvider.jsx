import React from 'react';
import { ThemeProvider as CssThemeProvider } from '@friendsofreactjs/react-css-themr';
import { createEmotionContextTheme } from './createEmotionContextTheme';

export const ThemeProvider = ({ theme, ...rest }) => (
  <CssThemeProvider theme={createEmotionContextTheme(theme)} {...rest} />
);
