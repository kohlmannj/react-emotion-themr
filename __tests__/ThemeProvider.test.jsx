import React from 'react';
import { shallow } from 'enzyme';
import { ThemeProvider } from '../src/ThemeProvider';

describe('ThemeProvider', () => {
  it('renders', () => {
    const wrapper = shallow(
      <ThemeProvider>
        <div />
      </ThemeProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
