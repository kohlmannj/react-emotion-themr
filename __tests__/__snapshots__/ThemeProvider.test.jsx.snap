// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider renders 1`] = `
<ThemeProvider
  theme={Object {}}
>
  <div />
</ThemeProvider>
`;
